modules/gapi/src/streaming/gstreamer/gstreamerptr.hpp

```diff
@@ -0,0 +1,248 @@
+{
+    if (pPtr)
+    {
+        gst_query_unref(*pPtr);
+        *pPtr = nullptr;
+    }
+}
+
+template<> inline void GStreamerPtrRelease<char>(char** pPtr)
+{
+    if (pPtr)
+    {
+        g_free(*pPtr);
+        *pPtr = nullptr;
+    }
+}
+
+// NOTE: The main concept of this class is to be owner of some passed to it piece of memory.
+//       (be owner = free this memory or reduce reference count to it after use).
+//       More specifically, GStreamerPtr is designed to own memory returned from GStreamer/GLib
+//       functions, which are marked as [transfer full] in documentation.
+//       [transfer full] means that function fully transfers ownership of returned memory to the
+//       receiving piece of code.
+//
+//       Memory ownership and ownership transfer concept:
+// https://developer.gnome.org/programming-guidelines/stable/memory-management.html.en#g-clear-object
+
+// NOTE: GStreamerPtr can only own strong references, not floating ones.
+//       For floating references please call g_object_ref_sink(reference) before wrapping
+//       it with GStreamerPtr.
+//       See https://developer.gnome.org/gobject/stable/gobject-The-Base-Object-Type.html#floating-ref
+//       for floating references.
+template<typename T>
+class GStreamerPtr
+{
+public:
+    inline GStreamerPtr() noexcept : ptr(nullptr) { }
+
+    explicit GStreamerPtr(T* p) noexcept : GStreamerPtr()
+    {
+        takeOwnership(p);
+    }
+
+    GStreamerPtr& operator=(T* p) noexcept
+    {
+        return takeOwnership(p);
+    }
+
+    T* get() { GAPI_Assert(ptr != nullptr); return ptr; }
+    // There is no const correctness in GStreamer C API
+    /*const*/ T* get() const { GAPI_Assert(ptr != nullptr); return (T*)ptr; }
+
+    T** getAddr() { GAPI_Assert(ptr == nullptr); return &ptr; }
+
+    inline operator T*() noexcept { return ptr; }
+    // There is no const correctness in GStreamer C API
+    inline operator /*const*/ T*() const noexcept { return (T*)ptr; }
+
+    const T* operator->() const { CV_Assert(ptr); return ptr; }
+
+    inline operator bool() const noexcept { return ptr != nullptr; }
+    inline bool operator!() const noexcept { return ptr == nullptr; }
+
+    inline T* releaseOwnership() noexcept
+    {
+        T* p = ptr;
+        ptr = nullptr;
+        return p;
+    }
+
+    inline void release() noexcept
+    {
+        if (ptr != nullptr)
+        {
+            GStreamerPtrRelease<T>(&ptr);
+        }
+    }
+
+    inline ~GStreamerPtr() noexcept
+    {
+        release();
+    }
+
+private:
+    // Pass result of functions with "transfer full" ownership
+    inline GStreamerPtr& takeOwnership(T* p) noexcept
+    {
+        release();
+        ptr = p;
+        return *this;
+    }
+
+    GStreamerPtr(const GStreamerPtr&) = delete;
+    GStreamerPtr(GStreamerPtr&&) = delete;
+    GStreamerPtr& operator=(const GStreamerPtr&) = delete;
+    GStreamerPtr& operator=(GStreamerPtr&&) = delete;
+
+private:
+    T* ptr;
```

Comment:
How's about to do not use home-grown implementation but default and highly-tested `std::unique_ptr` with custom dtor :)

```
tempate<...>
using GStreamerPtr = std::unique_ptr<Type, decltype(GStreamerPtrRelease<Type>)>

make_ptr(Type* ptr) {
return GStreamerPtr (ptr, &GStreamerPtrRelease<Type>);
}
```
}